src/imitation/algorithms/preference_comparisons.py

```diff
@@ -0,0 +1,289 @@
+            sampler=th.utils.data.RandomSampler(
+                dataset, replacement=True, num_samples=None
+            ),
+            collate_fn=preference_collate_fn,
+        )
+        self.optim.zero_grad()
+        for fragments, preferences in dataloader:
+            loss = self._loss(fragments, preferences)
+            loss.backward()
+        self.optim.step()
+
+
+class PreferenceComparisons:
+    """Main interface for reward learning using preference comparisons."""
+
+    def __init__(
+        self,
+        agent_trainer: AgentTrainer,
+        reward_model: RewardNet,
+        agent_timesteps: int,
+        fragmenter: Optional[Fragmenter] = None,
+        preference_gatherer: Optional[PreferenceGatherer] = None,
+        reward_trainer: Optional[RewardTrainer] = None,
+    ):
+        """Initialize the preference comparison trainer.
+
+        Args:
+            agent_trainer: generates trajectories while optionally training an RL agent
+                on the learned reward function (can also be a sampler from a static
+                dataset of trajectories though).
+            reward_model: a RewardNet instance to be used for learning the reward
+            agent_timesteps: number of environment steps to train the agent for between
+                each reward model training round
+            fragmenter: takes in a set of trajectories and returns pairs of fragments
+                for which preferences will be gathered. These fragments could be random,
+                or they could be selected more deliberately (active learning).
+                Default is a random fragmenter.
+            preference_gatherer: how to get preferences between trajectory fragments.
+                Default (and currently the only option) is to use synthetic preferences
+                based on ground-truth rewards. Human preferences could be implemented
+                here in the future.
+            reward_trainer: trains the reward model based on pairs of fragments and
+                associated preferences. Default is to use the preference model
+                and loss function from DRLHP.
+        """
+        self.model = reward_model
+        if reward_trainer is None:
+            reward_trainer = CrossEntropyRewardTrainer(reward_model)
+        self.reward_trainer = reward_trainer
+        self.agent = agent_trainer
+        if fragmenter is None:
+            fragmenter = RandomFragmenter()
+        self.fragmenter = fragmenter
+        if preference_gatherer is None:
+            preference_gatherer = SyntheticGatherer()
+        self.preference_gatherer = preference_gatherer
+        self.agent_timesteps = agent_timesteps
+
+        self.dataset = PreferenceDataset()
+
+    def train(self, steps: int):
+        """Train the reward model and the policy if applicable.
+
+        Args:
+            steps: number of iterations of the outer training loop
+        """
+        for _ in range(steps):
```

Comment:
Consider adding some logging here. You could do that in a separate PR, but I'd suggest doing that before trying to test whether your code works. Will be hard to debug without!